File: backend/app/calendar/schemas.py
from datetime import date, datetime
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class CalendarCategory(StrEnum):
    MACRO = "macro"
    EARNINGS = "earnings"
    DIVIDEND = "dividend"
    SPLIT = "split"
    CLOSED = "closed"


class CalendarScope(StrEnum):
    ALL = "all"
    WATCHLIST = "watchlist"
    HOLDING = "holding"


class CalendarDataSource(BaseModel):
    name: str
    state: Literal["ready", "stale", "error", "missing"]
    message: str | None = None


class CalendarEvent(BaseModel):
    id: str
    provider: str = "longbridge"
    provider_event_id: str | None = None
    category: CalendarCategory
    event_type: str | None = None
    title: str
    market: str | None = None
    country_or_region: str | None = None
    symbol: str | None = None
    security_name: str | None = None
    event_date: date
    event_datetime: datetime | None = None
    timezone: str = "Asia/Shanghai"
    all_day: bool = True
    financial_market_time: str | None = None
    importance: int | None = None
    period: str | None = None
    actual_value: str | None = None
    forecast_value: str | None = None
    previous_value: str | None = None
    revised_value: str | None = None
    unit: str | None = None
    currency: str | None = None
    content: str | None = None
    scope_tags: list[Literal["watchlist", "holding"]] = Field(default_factory=list)
    details: dict[str, Any] = Field(default_factory=dict)
    extra_data: dict[str, Any] = Field(default_factory=dict)
    source_name: str = "Longbridge"
    last_synced_at: datetime


class CalendarEventGroup(BaseModel):
    event_date: date
    items: list[CalendarEvent]


class CalendarEventsResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    category: CalendarCategory
    from_date: date = Field(alias="from")
    to_date: date = Field(alias="to")
    timezone: str = "Asia/Shanghai"
    items: list[CalendarEvent]
    groups: list[CalendarEventGroup]
    next_cursor: str | None = None
    last_synced_at: datetime | None = None
    data_source: CalendarDataSource


class CalendarFiltersResponse(BaseModel):
    category: CalendarCategory
    markets: list[str]
    importance: list[int]
    scopes: list[CalendarScope]


class CalendarRefreshRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    category: CalendarCategory
    from_date: date = Field(alias="from")
    to_date: date = Field(alias="to")
    markets: list[str] = Field(default_factory=list)
    scope: CalendarScope = CalendarScope.ALL
    importance: list[int] = Field(default_factory=list)

File: frontend/src/features/calendar/schemas.ts
import { z } from "zod";

const categorySchema = z.enum(["macro", "earnings", "dividend", "split", "closed"]);

const calendarEventSchema = z
  .object({
    id: z.string(),
    provider: z.string(),
    provider_event_id: z.string().nullable(),
    category: categorySchema,
    event_type: z.string().nullable(),
    title: z.string(),
    market: z.string().nullable(),
    country_or_region: z.string().nullable(),
    symbol: z.string().nullable(),
    security_name: z.string().nullable(),
    event_date: z.string(),
    event_datetime: z.string().nullable(),
    timezone: z.string(),
    all_day: z.boolean(),
    financial_market_time: z.string().nullable(),
    importance: z.number().nullable(),
    period: z.string().nullable(),
    actual_value: z.string().nullable(),
    forecast_value: z.string().nullable(),
    previous_value: z.string().nullable(),
    revised_value: z.string().nullable(),
    unit: z.string().nullable(),
    currency: z.string().nullable(),
    content: z.string().nullable(),
    scope_tags: z.array(z.enum(["watchlist", "holding"])),
    details: z.record(z.string(), z.union([z.string(), z.number(), z.null()])),
    extra_data: z.record(z.string(), z.unknown()),
    source_name: z.string(),
    last_synced_at: z.string(),
  })
  .passthrough();

export const calendarEventsSchema = z
  .object({
    category: categorySchema,
    from: z.string(),
    to: z.string(),
    timezone: z.string(),
    items: z.array(calendarEventSchema),
    groups: z.array(
      z
        .object({
          event_date: z.string(),
          items: z.array(calendarEventSchema),
        })
        .passthrough(),
    ),
    next_cursor: z.string().nullable(),
    last_synced_at: z.string().nullable(),
    data_source: z
      .object({
        name: z.string(),
        state: z.enum(["ready", "stale", "error", "missing"]),
        message: z.string().nullable(),
      })
      .passthrough(),
  })
  .passthrough();

export const calendarFiltersSchema = z
  .object({
    category: categorySchema,
    markets: z.array(z.string()),
    importance: z.array(z.number()),
    scopes: z.array(z.enum(["all", "watchlist", "holding"])),
  })
  .passthrough();
